sdk/metric/controller/basic/controller.go

```diff
@@ -0,0 +1,302 @@
+// controller was already started.
+//
+// Note that it is not necessary to Start a controller when only
+// pulling data; use the Collect() and ForEach() methods directly in
+// this case.
+func (c *Controller) Start(ctx context.Context) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.ticker != nil {
+		return ErrControllerStarted
+	}
+
+	c.wg.Add(1)
+	c.ticker = c.clock.Ticker(c.period)
+	go c.runTicker(ctx, c.stopCh)
+	return nil
+}
+
+// Stop waits for the background goroutine to return and then collects
+// and exports metrics one last time before returning.  The passed
+// context is passed to the final Collect() and subsequently to the
+// final asynchronous instruments.
+//
+// Note that Stop() will not cancel an ongoing collection or export.
+func (c *Controller) Stop(ctx context.Context) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.stopCh == nil {
+		return nil
+	}
+
+	close(c.stopCh)
+	c.stopCh = nil
+	c.wg.Wait()
+	c.ticker.Stop()
+
+	return c.collect(ctx)
+}
+
+// runTicker collection on ticker events until the stop channel is closed.
+func (c *Controller) runTicker(ctx context.Context, stopCh chan struct{}) {
+	defer c.wg.Done()
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-c.ticker.C():
+			if err := c.collect(ctx); err != nil {
+				otel.Handle(err)
+			}
+		}
+	}
+}
+
+// collect computes a checkpoint and optionally exports it.
+func (c *Controller) collect(ctx context.Context) error {
+	if err := c.checkpoint(ctx, func() bool {
+		return true
+	}); err != nil {
+		return err
+	}
+	if c.exporter == nil {
+		return nil
+	}
+	// This export has to complete before starting another
+	// collection since it will hold a read lock and
+	// checkpoint() must re-acquire a write lock.
+	if err := c.export(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// checkpoint calls the Accumulator and Checkpointer interfaces to
+// compute the CheckpointSet.  This applies the configured collection
+// timeout.  Note that this does not try to cancel a Collect or Export
+// when Stop() is called.
+func (c *Controller) checkpoint(ctx context.Context, cond func() bool) error {
+	ckpt := c.checkpointer.CheckpointSet()
+	ckpt.Lock()
+	defer ckpt.Unlock()
+
+	if !cond() {
+		return nil
+	}
+	c.checkpointer.StartCollection()
+
+	if c.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	_ = c.accumulator.Collect(ctx)
+
+	var err error
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+	default:
+		// The context wasn't done, ok.
+	}
+
+	// Finish the checkpoint whether the accumulator timed out or not.
+	if cerr := c.checkpointer.FinishCollection(); cerr != nil {
+		if err == nil {
+			err = cerr
+		} else {
+			err = fmt.Errorf("%s: %w", cerr.Error(), err)
+		}
+	}
+
+	return err
+}
+
+// export calls the exporter with a read lock on the CheckpointSet,
+// applying the configured export timeout.
+func (c *Controller) export() error {
```

Comment:
should we pass context as an argument here instead of creating a context.Background below?

Reply:
Fixed.